File: confmaster/src/main/java/com/navercorp/nbasearc/confmaster/logger/Log.java
/*
 * Copyright 2015 Naver Corp.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.navercorp.nbasearc.confmaster.logger;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.io.Writer;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;

import org.apache.log4j.Level;

public class Log {
    
    private static final DateFormat df = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss.SSS");
    private final Level level;
    private final Date time;
    private final String msg;
    private final Throwable e;

    public Log(Level level, Date time, String msg) {
        this.level = level;
        this.time = time;
        this.msg = msg;
        this.e = null;
    }

    public Log(Level level, Date time, String msg, Throwable e) {
        this.level = level;
        this.time = time;
        this.msg = msg;
        this.e = e;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        
        sb.append(
            String.format("%s %s %03d %s", 
                df.format(time), getLevel().toString().substring(0, 3), 
                Thread.currentThread().getId(), msg));
        if (e != null) {
            sb.append(" ").append(getStackTrace(e));
        }
        
        return sb.toString();
    }

    public static String getStackTrace(Throwable aThrowable) {
      Writer result = new StringWriter();
      PrintWriter printWriter = new PrintWriter(result);
      aThrowable.printStackTrace(printWriter);
      return result.toString();
    }

    public Level getLevel() {
        return level;
    }
    
    public boolean hasException() {
        return e != null;
    }

}

File: smr/test/integrated/Log.py
#
# Copyright 2015 Naver Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import os
import subprocess
import shlex
import Util, Conf

Logs = set()

def createlog(path):
    if Conf.USE_MEM_LOG:
        os.system('%s createlog %s' % (Conf.LOG_UTIL_BIN_PATH, path))
        Logs.add(path)

def syncdeletelog(path):
    if path in Logs:
        os.system('%s synclog %s' % (Conf.LOG_UTIL_BIN_PATH, path))
        os.system('%s deletelog %s' % (Conf.LOG_UTIL_BIN_PATH, path))
        Logs.discard(path)

def atExit():
    while len(Logs) > 0:
        path = Logs.pop()
        os.system('%s deletelog %s' % (Conf.LOG_UTIL_BIN_PATH, path))

def communicate(cmd):
    popen = subprocess.Popen(shlex.split(cmd), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    return popen.communicate()

def datadump_communicate(path, begin_seq, end_seq = None, oflag = 'Tshld'):
    prog = Conf.LOG_UTIL_BIN_PATH
    if end_seq != None:
        cmd = '%s -o %s datadump %s %d %d' % (prog, oflag, path, begin_seq, end_seq)
    else:
        cmd = '%s -o %s datadump %s %d' % (prog, oflag, path, begin_seq)
    return communicate(cmd)

def dumpbytime_communicate(path, begin_time, end_time = None, oflag = 'Tshld'):
    prog = Conf.LOG_UTIL_BIN_PATH
    if end_time != None:
        cmd = '%s -o %s dumpbytime %s %d %d' % (prog, oflag, path, begin_time, end_time)
    else:
        cmd = '%s -o %s dumpbytime %s %d' % (prog, oflag, path, begin_time)
    return communicate(cmd)

def mincorelog_communicate (path, seq = None):
    prog = Conf.LOG_UTIL_BIN_PATH
    if seq != None:
        cmd = '%s mincorelog %s %d' % (prog, path, seq)
    else:
        cmd = '%s mincorelog %s' % (prog, path)
    return communicate(cmd)

def decachelog_communicate (path, seq = None, force = False):
    prog = Conf.LOG_UTIL_BIN_PATH

    if seq == None:
        cmd = '%s decachelog %s' % (prog, path)
    else:
        if force == True:
            suffix = " force"
        else:
            suffix = ""
        cmd = '%s decachelog %s %d%s' % (prog, path, seq, suffix)

    return communicate(cmd)
